### PR TITLE
Add pre and post snapshot and prune hooks

### LINKED ISSUE #5: Creating a snapshot of a snapshot
Add an option to create a snapshot of a snapshot into the right click selection menu.

### LINKED ISSUE #12: very confusing
### Describe the problem your feature would solve
the initial setup is quite confusing for me. 
Ok, first I press the plus button. Then I give it a name for the snapshot. So far so good. But I'm lost where I should set the snapshot location or the subvolume. What is a subvolume and what has it to do with btrfs? Should I locate the snapsots in root or userland?
I have no Idea so I leave everything default but have no Idea if I'm backing up the whole root files or only root/var/lib/machines (as left default).

Next I want to set a hourly schedule. So I find "Retention Policy" ok, I can guess what that is and I leave it default also, because I'm not quite sure what it really exactly is. So now I click on Subvolumes. I have to add another subvolume? Didn't I just add one on the main screen? What did I add on the main screen for? Will this schedule that I'm adding right now be a new one or for the subvolume that I already added on the main screen?
Now the next question is I want to snapshot only root files not the home folder. How would I do this? I couldn't find out!
The next question would be where should I save the snapshots? Can I save them anywhere or is it better so save them in some designated folder?
Sorry, had to give up.

### Describe the solution you'd like
Let the software choose where to store snapshots and decide between root files and/or home folder

### LINKED ISSUE #15: Post snapshot hooks
### Describe the problem your feature would solve
Be able to take actions after a snapshot like clean up or move backup to offsite etc

### Describe the solution you'd like
Could you add an option to execute scripts on state changes like when a snapshot completes

### Additional context
Post / Pre operation hooks is a fairly common feature in most backup suites and even though I love Butter's simplicity, just having this feature would allow for a lot more complex workflows for those who want it.
Personally I want to use this for sending the btrfs snapshot to a local server to have backups be available wven if the device fails,
there are a lot of BTRFS backup solutions but seems like none of them have gotten around to or have any ability to backup to a remote service, which I think is fairly critical

### PR DESCRIPTION
**Pull request by sidevesh**
- [x] I have read and agree to the Contributing Guidelines.
- [x] This compiles and is ready for review.

Closes #15

### CODE REVIEW

**Comment by zhangyuannie** [comment]
Thanks for the amazing work! I'm quite busy recently unfortunately, will hopefully take a look soon (ETA >= 2w).

**Comment by sidevesh** [comment]
@zhangyuannie were you able to take a look ?

File: src/rule.rs [hunk 1]
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
...
 
 impl RuleSubvolume {
-    fn snapshot(&self) -> anyhow::Result<()> {
+    fn snapshot(&self, rule: &Rule) -> anyhow::Result<()> {
         let mut name = RandomName::new();
         for _ in 0..16 {
             let target_path = self.target_dir.join(name.as_str());
+            if !rule.pre_snapshot.is_empty() {
+                let mut cmd = std::process::Command::new("sh");
+                cmd.arg("-c").arg(format!("{} {} {}", rule.pre_snapshot, self.path.display(), target_path.display()));
```
Comment (zhangyuannie) [inline src/rule.rs:162]: We probably also need to document this somewhere. We probably want to also have an option for letting this script run in root/user mode. For now I think you can do a simple check to make sure this file is only writable by root

File: src/rule.rs [hunk 2]
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
...
+                match res {
+                    Ok(output) => {
+                        if !output.status.success() {
+                            let error_code = output.status.code();
+                            if let Some(code) = error_code {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, exit code: {}, error: {}", code, String::from_utf8_lossy(&output.stderr));
+                            } else {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, error: {}", String::from_utf8_lossy(&output.stderr));
+                            }
+                            return Err(anyhow::anyhow!("Pre snapshot script failed"));
```
Comment (zhangyuannie) [inline src/rule.rs:173]: should we abort snapshotting because user's script failed?

**Review by zhangyuannie — COMMENTED** [review]
I started looking at this last week but wasn't able to finish. There's quite a lot of details.

TLDR: I'm ok shipping this if:

- We change the json schema from `pre_snapshot` to `pre_snapshot_EXPERIMENTAL`, etc.
- We at least make sure the script we ran is only writable by root (or run it as user instead)
- Maybe not showing on the UI yet (or just the post snapshot hook to start with)

File: src/rule.rs [hunk 3]
```diff
@@ -232,12 +284,43 @@ impl RuleSubvolume {
             }
 
             if should_remove {
+                if !rule.pre_prune.is_empty() {
```
[the remark this answers is not in the document]
Comment (sidevesh) [reply src/rule.rs:287]: I agree, pre / post prune is gonna be very useless for most people,
my use case for it is to cleanup snpshots on the external backup once its pruned from the system,
there maybe other cleanup related stuff that user might want to do when a snapshot is pruned but tbh I can't think of any other usecase either.

File: src/rule.rs (hunk 1 above)
Comment (sidevesh) [reply src/rule.rs:162]: Is the check if the script is only writable by root so that someone without root can't use it for executing things as root ?
That would be an issue though even if you meant the check for something else, I will add that.

File: src/rule.rs (hunk 2 above)
Comment (sidevesh) [reply src/rule.rs:173]: my usecase here is to use the pre snapshot to check / mount the external backup location,
local snapshots are useless for me as a backup thing if it is not kept externally,
using a pre action script to check things before letting action happen is a pattern I have seen and could be useful for other usecases.
With btrfs snapshots being so cheap, there isnt really much difference otherwise between pre and post, except that pre can prevent backup from happening if so desired.

File: src/rule.rs [hunk 4]
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
...
+                        return Err(e.into());
+                    }
+                }
+            }
             match create_butter_snapshot(&self.path, &target_path, true) {
-                Ok(_) => return Ok(()),
+                Ok(_) => {
+                    if !rule.post_snapshot.is_empty() {
+                        let mut cmd = std::process::Command::new("sh");
+                        cmd.arg("-c").arg(format!("{} {} {}", rule.post_snapshot, self.path.display(), target_path.display()));
```
[the remark this answers is not in the document]
Comment (sidevesh) [reply src/rule.rs:186]: Post snapshot hook here should be getting the path to the snapshot.
Taking custom command with support for placeholders for the variables we insert could be useful for quick things but I don't think it would make much difference right now.

File: data/resources/ui/schedule_rule_edit_dialog.ui [hunk 5]
```diff
@@ -192,6 +192,74 @@
...
                 </child>
 
+                <child>
+                  <object class="AdwPreferencesGroup">
+                    <property name="title" translatable="yes">Pre and Post Snapshot scripts</property>
+                    <property name="description" translatable="yes">Scripts to run before and after the snapshot,&#10;Scripts get the source and target paths passed as arguments,&#10;If the Pre Snapshot script is provided and it fails then the backup would not be performed,&#10;Please note that the scripts are run as root and you may cause damage to your system if you are not careful.</property>
+
+                    <child>
+                      <object class="AdwActionRow">
+                        <property name="title" translatable="yes">Pre Snapshot</property>
```
[the remark this answers is not in the document]
Comment (sidevesh) [reply data/resources/ui/schedule_rule_edit_dialog.ui:202]: I agree, my goal with this PR was to experiment and try out ways to move and manage backups on external location,
and that needed all the 4 script hooks that I added.
I am trying to model my setup similar to Time Machine on macOS and its backup and cleanup scheme for snapshots and external backups.
And I have realized that a backup tool like that does not fit Butter's workflow exactly,
I think making the pre / post prune fields experimental and not showing them in UI right now would make sense right now,
I think we should keep the pre snapshot as it might be useful in some usecases and pre / post hooks around an action is a common pattern in general but also in other backup tools.

File: src/rule.rs (hunk 1 above)
Comment (zhangyuannie) [reply src/rule.rs:162]: It's so that someone without root can't modify the script.

**Comment by sidevesh** [comment]
@zhangyuannie are you okay with accepting this with following changes:
  -- Only run scripts if they are only writable by root, otherwise fail, add to the message in section subtitle informing about this
  -- Make pre / post prune hooks `_EXPERIMENTAL` in json schema and hide it from UI

**Comment by zhangyuannie** [comment]
Please also add _EXPERIMENTAL to pre and post snapshot hooks. They can show up on the UI if its label also mention that this feature is experimental.

This is to avoid commiting to something that we will regret. For example, you mentioned:

> Post snapshot hook here should be getting the path to the snapshot.

However, in your current code, that is not the case.

**Comment by zhangyuannie** [comment]
Hi @sidevesh, are you still interested in moving this forward?

**Comment by sidevesh** [comment]
hey @zhangyuannie , I am but I wasn't able to clean this up due to losing access to my linux laptop while travelling.
I will look into addressing the points in next weeks sometime once I am back with my laptop.

**Comment by sidevesh** [comment]
I have had quite a bit of success with making Butter move snapshots to a sparsefile kept in samba share using these pre post snapshot hooks,
I was working on moving that process out from the script that I assign to the hooks to Butter's rust code itself,
allowing the user to specify the network share details and location to keep the sparsefile disk image on and then moving the snapshot after the backup.
So my original need for hooks is now solved in my wip fork of Butter, still makes sense to get atleast pre post snapshot hooks  in as that's anyways a standard thing in most backup solutions.

**Comment by sidevesh** [comment]
Question, would you be interested in upstreaming the network backup functionality I am working on in my fork into Butter itself ?

**Comment by zhangyuannie** [comment]
> allowing the user to specify the network share details and location to keep the sparsefile disk image on and then moving the snapshot after the backup.

I have very limited experience with sparsefile and samba so would need to understand how your solution works. So ideally I can have a section of "community supported features/extensions" and just dump it there but probably not gonna happen anytime soon if at all. :stuck_out_tongue_closed_eyes:

So I would say let's have pre/post snapshot/prune hooks first with `_EXPERIMENTAL` suffix everywhere. Thanks!
